neps/nep-0508.md

```diff
@@ -0,0 +1,264 @@
+
+* Merging shards together
+* Moving the boundary account between two shards
+
+### Localization of resharding event to specific shard
+
+As of today, at the RocksDB storage layer, we have the ShardUId, i.e. the ShardId along with the ShardVersion, as a prefix in the key of trie state and flat state. During a resharding event, we increment the ShardVersion by one, and effectively remap all the current parent shards to new child shards. This implies we can't use the same underlying key value pairs for store and instead would need to duplicate the values with the new ShardUId prefix, even if a shard is unaffected and not split.
+
+In the future, we would like to potentially change the schema in a way such that only the shard that is splitting is impacted by a resharding event, so as to avoid additonal work done by nodes tracking other shards.
+
+### Other useful features
+
+* Removal of shard uids and introducing globally unique shard ids
+* Account colocation for low latency across account call - In case we start considering synchronous execution environment, colocating associated accounts (e.g. cross contract call between them) in the same shard can increase the efficiency 
+* Shard purchase/reservation - When someone wants to secure entirety of limitation on a single shard (e.g. state size limit), they can 'purchase/reserve' a shard so it can be dedicated for them (similar to how Aurora is set up)
+
+## Consequences
+
+### Positive
+
+* Workload across shards will be more evenly distributed.
+* Required space to maintain state (either in memory or in persistent disk) will be smaller. This is useful for in-memory tries.
+* State sync overhead will be smaller with smaller state size.
+
+### Neutral
+
+* Number of shards would increase.
+* Underlying trie structure and data structure are not going to change.
+* Resharding will create dependency on flat state snapshots.
+* The resharding process, as of now, is not fully automated. Analyzing shard data, determining the split boundary, and triggering an actual shard split all need to be manually curated and tracked.
+
+### Negative
+
+* During resharding, a node is expected to require more resources as it will first need to copy state data from the parent shard to the child shard, and then will have to apply trie and flat state changes twice, once for the parent shard and once for the child shards.
+* Increased potential for apps and tools to break without proper shard layout change handling.
+
+### Backwards Compatibility
+
+Any light clients, tooling or frameworks external to nearcore that have the current shard layout or the current number of shards hardcoded may break and will need to be adjusted in advance. The recommended way for fixing it is querying an RPC node for the shard layout of the relevant epoch and using that information in place of the previously hardcoded shard layout or number of shards. 
```

Comment:
It might be nice to show an example of this query.

As far as I know, the main reason to have a hard-coded shard id is to query transactions from a specific account by downloading just the chunk that pertains to that account. This can be done using the `block_id` + `shard_id` version of the [chunk details query](https://docs.near.org/api/rpc/block-chunk#chunk-details).

After a split, where an account might move to a larger shard id, how should a tool recover? How will the tool know that it needs to change which shard id it is querying? If the tool is simply iterating over the chunk looking for a specific account, it will not be able to distinguish a chunk with happens to not including any transactions from that account from the situation where a split just happened and there are transactions now located in a different chunk?

Reply:
Thanks! Can you share an example of an implementation of this pattern? The adjusted pattern may look something like the following. I'll implement the new RPC endpoints soon and update the NEP to reflect that. 
- shard_layout := rpc(method="shard_layout", block_hash=block_hash)
- shard_id = shard_layout.get_shard_id_for_account(account)
- ... the rest same as before

This would be the least intrusive change but frankly quering the whole chunk just to get transactions for a single account could be done with a dedicated rpc endpoint. This is outside of the scope of this NEP though and we'll focus on delivering the relavant endpoints.

Reply:
I don't know if there is a production implementation of this pattern somewhere, but this kind of query is something I have done myself in the past when looking at recent transactions coming from `relay.aurora` for debugging purposes. The pseudocode looked something like

```
AURORA_RELAYER_SHARD_ID = 3

block_height = read_from_stdin()
chunk = rpc(method="chunk", block_id=block_height, shard_id=AURORA_RELAYER_SHARD_ID)
txs = [t for t in chunk.transactions if t.signer_id == "relay.aurora"]
print(txs)
```

Your proposed RPC method where the shard layout can be directly queried would allow getting rid of the hard-coded shard id and I agree it is the least intrusive.

Reply:
Added instructions on how to get the current shard layout by using the rpc endpoint EXPERIMENTAL_protocol_config.